### PR TITLE
IOTEDGE-1012 Add OAuth 2.0 Device Authorization Grant methods to SDK

### PR DESCRIPTION
**Pull request by JacoJooste**

### CODE REVIEW

File: tests/internal/anvil/am/amclient.go [hunk 1]
```diff
@@ -583,3 +595,46 @@ func SetAdvancedServerProperties(properties map[string]interface{}) (err error)
...
+	}
+	// the response is an HTML page with an embedded JSON object that contains the pageData message
+	// we need to extract the JSON and parse it to read the pageData
+	re := regexp.MustCompile("(?s:pageData.*})")
+	pageData := re.FindString(string(responseBodyBytes))
+	re = regexp.MustCompile("(errorCode:\\s*\")(.*)(\")")
+	if re.MatchString(pageData) {
+		return fmt.Errorf("request failed with error code: " + re.FindStringSubmatch(pageData)[2])
+	}
+	re = regexp.MustCompile("(done:\\s*)(.*)")
```
Comment (JacoJooste) [inline tests/internal/anvil/am/amclient.go:635]:
```suggestion
	re = regexp.MustCompile(`(done:\s*)(.*)`)
```

File: tests/internal/anvil/am/amclient.go [hunk 2]
```diff
@@ -583,3 +595,46 @@ func SetAdvancedServerProperties(properties map[string]interface{}) (err error)
...
+		return err
+	}
+	if response.StatusCode != 200 {
+		return fmt.Errorf("user consent response: %s", string(responseBodyBytes))
+	}
+	// the response is an HTML page with an embedded JSON object that contains the pageData message
+	// we need to extract the JSON and parse it to read the pageData
+	re := regexp.MustCompile("(?s:pageData.*})")
+	pageData := re.FindString(string(responseBodyBytes))
+	re = regexp.MustCompile("(errorCode:\\s*\")(.*)(\")")
```
Comment (JacoJooste) [inline tests/internal/anvil/am/amclient.go:631]:
```suggestion
	re = regexp.MustCompile(`(errorCode:\s*")(.*)(")`)
```

File: tests/internal/anvil/am/amclient.go [hunk 3]
```diff
@@ -583,3 +595,46 @@ func SetAdvancedServerProperties(properties map[string]interface{}) (err error)
...
+	responseBodyBytes, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return err
+	}
+	if response.StatusCode != 200 {
+		return fmt.Errorf("user consent response: %s", string(responseBodyBytes))
+	}
+	// the response is an HTML page with an embedded JSON object that contains the pageData message
+	// we need to extract the JSON and parse it to read the pageData
+	re := regexp.MustCompile("(?s:pageData.*})")
```
Comment (JacoJooste) [inline tests/internal/anvil/am/amclient.go:629]:
```suggestion
	re := regexp.MustCompile(`(?s:pageData.*})`)
```

File: internal/gateway/gateway_test.go [hunk 4]
```diff
@@ -29,75 +29,17 @@ import (
 	"testing"
 	"time"
 
+	"github.com/ForgeRock/iot-edge/v7/internal/mocks"
+
```
Comment (LimaEchoCharlie) [inline internal/gateway/gateway_test.go:33]: Is this gap intentional?

File: internal/thing/thing.go [hunk 5]
```diff
@@ -37,6 +37,16 @@ import (
 	"gopkg.in/square/go-jose.v2/jwt"
 )
 
+const (
+	// Device authorization grant constants: https://tools.ietf.org/html/rfc8628#section-3.5
```
Comment (LimaEchoCharlie) [inline internal/thing/thing.go:41]: 🥇

File: internal/gateway/gateway_test.go (hunk 4 above)
Comment (JacoJooste) [reply internal/gateway/gateway_test.go:33]: No, it's just a result of running two different import sorting types. I'll fix it up.

File: tests/internal/anvil/am/amclient.go [hunk 6]
```diff
@@ -25,9 +25,13 @@ import (
 	"log"
 	"net/http"
 	"net/http/httputil"
+	"net/url"
+	"regexp"
 	"strings"
 	"time"
 
+	"github.com/ForgeRock/iot-edge/v7/pkg/thing"
+
```
Comment (LimaEchoCharlie) [inline tests/internal/anvil/am/amclient.go:34]: Same gap.

File: tests/internal/anvil/am/amclient.go [hunk 7]
```diff
@@ -239,14 +243,23 @@ func get(endpoint string, version string) (reply []byte, err error) {
...
 func getSSOToken() (token string, err error) {
-	req, err := http.NewRequest(http.MethodPost, AMURL+"/json/realms/root/authenticate", nil)
+	attributes := IdAttributes{
+		Name:     adminUsername,
+		Password: adminPassword,
+	}
+	return getSSOTokenForIdentity("/", attributes)
+}
+
+// getSSOTokenForIdentity gets an SSO token using the AM Admin's credentials
```
Comment (LimaEchoCharlie) [inline tests/internal/anvil/am/amclient.go:253]: `AM Admin` -> `User`

File: tests/internal/anvil/am/amclient.go [hunk 8]
```diff
@@ -583,3 +596,46 @@ func SetAdvancedServerProperties(properties map[string]interface{}) (err error)
...
+	response, err := httpClient.Do(request)
+	if err != nil {
+		return err
+	}
+	defer response.Body.Close()
+	responseBodyBytes, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return err
+	}
+	if response.StatusCode != 200 {
```
Comment (LimaEchoCharlie) [inline tests/internal/anvil/am/amclient.go:625]: `200` -> `http.StatusOK`

**Review by LimaEchoCharlie — APPROVED** [review]
Awesome work Jaco!

**Pull request merged**